File: ecobertura.core/src/ecobertura/core/launching/JavaApplicationLaunchDelegate.java
package ecobertura.core.launching;

import org.eclipse.core.runtime.CoreException;
import org.eclipse.core.runtime.IConfigurationElement;
import org.eclipse.core.runtime.IExecutableExtension;
import org.eclipse.core.runtime.IProgressMonitor;
import org.eclipse.debug.core.ILaunch;
import org.eclipse.debug.core.ILaunchConfiguration;
import org.eclipse.debug.core.ILaunchManager;
import org.eclipse.debug.core.model.ILaunchConfigurationDelegate2;

import ecobertura.core.CorePlugin;

public class JavaApplicationLaunchDelegate implements
		ILaunchConfigurationDelegate2, IExecutableExtension {

	private ILaunchConfigurationDelegate2 delegateToExtend; 
	
	@Override
	public void setInitializationData(IConfigurationElement config,
			String propertyName, Object data) throws CoreException {
		delegateToExtend = LaunchDelegateRetriever.fromExtensionConfig(config).getDelegate();
	}
	
	@Override
	public void launch(ILaunchConfiguration configuration, String mode,
			ILaunch launch, IProgressMonitor monitor) throws CoreException {
		
	    // TODO monitor
		
		final LaunchInstrumenter launchInstrumenter = 
			LaunchInstrumenter.instrumentClassesFor(configuration);
		final ILaunchConfiguration updatedConfiguration = 
			launchInstrumenter.getUpdatedLaunchConfiguration();
		CorePlugin.instance().coverageResultsCollector().coveredLaunchStarted(updatedConfiguration);
		delegateToExtend.launch(updatedConfiguration, ILaunchManager.RUN_MODE, launch, monitor);
	}

	@Override
	public boolean buildForLaunch(ILaunchConfiguration configuration,
			String mode, IProgressMonitor monitor) throws CoreException {

		// simply forward for now
		return delegateToExtend.buildForLaunch(configuration, mode, monitor);
	}

	@Override
	public boolean finalLaunchCheck(ILaunchConfiguration configuration,
			String mode, IProgressMonitor monitor) throws CoreException {
		
		// simply forward for now
		return delegateToExtend.finalLaunchCheck(configuration, mode, monitor);
	}

	@Override
	public ILaunch getLaunch(ILaunchConfiguration configuration, String mode)
			throws CoreException {
		
		// simply forward for now
		return delegateToExtend.getLaunch(configuration, mode);
	}

	@Override
	public boolean preLaunchCheck(ILaunchConfiguration configuration,
			String mode, IProgressMonitor monitor) throws CoreException {
		
		// simply forward for now
		return delegateToExtend.preLaunchCheck(configuration, mode, monitor);
	}
}

File: ecobertura.core/src/ecobertura/core/launching/JavaApplicationLaunchDelegate.scala
package ecobertura.core.launching

import org.eclipse.core.runtime._
import org.eclipse.debug.core._
import org.eclipse.debug.core.model.ILaunchConfigurationDelegate2

import _root_.ecobertura.core.CorePlugin

class JavaApplicationLaunchDelegate extends ILaunchConfigurationDelegate2 
		with IExecutableExtension {

	private var delegateToExtend: ILaunchConfigurationDelegate2 = null
	
	override def setInitializationData(config: IConfigurationElement, propertyName: String, data: Any) =
		delegateToExtend = LaunchDelegateRetriever.fromExtensionConfig(config).getDelegate
	
	override def launch(configuration: ILaunchConfiguration, mode: String, launch: ILaunch, 
			monitor: IProgressMonitor) = {
		
		// TODO monitor
		
		val launchInstrumenter = LaunchInstrumenter instrumentClassesFor configuration
		val updatedConfiguration = launchInstrumenter.getUpdatedLaunchConfiguration
		
		CorePlugin.instance.coverageResultsCollector coveredLaunchStarted launch
		delegateToExtend launch (updatedConfiguration, ILaunchManager.RUN_MODE, launch, monitor)
	}
	
	override def buildForLaunch(configuration: ILaunchConfiguration, mode: String,
			monitor: IProgressMonitor) = {
		// simply forward for now
		delegateToExtend buildForLaunch (configuration, mode, monitor)
	}
	
	override def finalLaunchCheck(configuration: ILaunchConfiguration, mode: String,
			monitor: IProgressMonitor) = {
		// simply forward for now
		delegateToExtend finalLaunchCheck (configuration, mode, monitor)
	}
	
	override def getLaunch(configuration: ILaunchConfiguration, mode: String) = {
		// simply forward for now
		delegateToExtend getLaunch (configuration, mode)
	}

	override def preLaunchCheck(configuration: ILaunchConfiguration, mode: String,
			monitor: IProgressMonitor) = {
		// simply forward for now
		delegateToExtend preLaunchCheck (configuration, mode, monitor)
	}
}
